# Lab 05 - Ufuncs & Pandas - Andrew Badzioch

- What are NumPy Universal Functions (ufuncs)
1. The NumPy ufuncs are functions operating on NumPy arrays.
2. The ufuncs are known as vectorized operations that make repeated calculations on arrays faster and are more effective.
3. Vectorized operations perform better on vectors rather than loops. This is why unfuncs are not performing iterations.

In [1]:
# an example using the timeit library

# importing timeit
import timeit

In [2]:
# 1. create two python lists
a = [4, 8, 3, 2]
b = [1, 6, 7, 9]

In [3]:
# 2. add these lists with normal python operations
c = a + b
print(c)
# let's time the operation
%timeit c

[4, 8, 3, 2, 1, 6, 7, 9]
6.23 ns ± 0.00492 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)


In [4]:
# 3. using numpy ufuncs adding a and b
import numpy as np
d = np.add(a, b)

print(d)
%timeit d

[ 5 14 10 11]
6.03 ns ± 0.0457 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)


In [5]:
e = np.array([2,3,4,5,6])

In [6]:
print(e+5)

[ 7  8  9 10 11]


In [7]:
e/2

array([1. , 1.5, 2. , 2.5, 3. ])

In [8]:
print(e*3)

[ 6  9 12 15 18]


In [9]:
e*3

array([ 6,  9, 12, 15, 18])

In [10]:
e//3

array([0, 1, 1, 1, 2])

In [11]:
print(np.multiply.reduce(e))

720


### Lab 05 - Pandas Part 2 - Andrew Badzioch (beginning of the data analysis fun)
- You need to add observations after every output that provides a new piece of knowledge about the dataset.

##### Steps of a DS project
1. Introduction (Background, Objectives, Data Description) (variables=description, )
2. Install and import libraries
3. Load (read) the dataset (always create a copy of the dataset)
4. Basic data exploration
5. EDA
 - data cleaning
 - data engineering
 - data visualization
 - handling missing valuse
 - correlation
 - skewness
 - EDA summary
6. Data mining
7. Conclusion (presenting your findings and providing actionable insights)

#### 1. Introduction 
- complete  all 3 parts of the introduction for this dataset
-  The movies datset contains 
- to practice 
- 

#### 2. Importing libraries

In [12]:
# general libraries
import numpy as np
import pandas as pd
import math

# visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

# suppressing library (to suppress warnings)
import warnings
warnings.filterwarnings('ignore')

#### 3. Loading (reading) the dataset

In [13]:
# let's load the dataset
df = pd.read_csv('movies.csv')

In [14]:
# create a copy of the file
movies = df.copy()

### 4. Basic data exploration

In [15]:
# basic data exploration with head(): methos to view the first 5 rows
movies.head()

Rank                         Title       Studio       Gross  Year
0     1             Avengers: Endgame  Buena Vista  $2,796.30   2019
1     2                        Avatar          Fox  $2,789.70   2009
2     3                       Titanic    Paramount  $2,187.50   1997
3     4  Star Wars: The Force Awakens  Buena Vista  $2,068.20   2015
4     5        Avengers: Infinity War  Buena Vista  $2,048.40   2018

In [16]:
movies.tail()

Rank                     Title           Studio     Gross  Year
777   778                 Yogi Bear  Warner Brothers  $201.60   2010
778   779       Garfield: The Movie              Fox  $200.80   2004
779   780               Cats & Dogs  Warner Brothers  $200.70   2001
780   781  The Hunt for Red October        Paramount  $200.50   1990
781   782                  Valkyrie              MGM  $200.30   2008

#### Data attributes:
1. shape:
2. size:
3. ndim:
4. dtypes:
5. columns:

In [17]:
movies.shape

(782, 5)

Observation
- there are 782 rows and 5 columns in this dataset

In [18]:
movies.size

3910

In [19]:
movies.ndim

2

In [20]:
movies.dtypes # variables

Rank       int64
Title     object
Studio    object
Gross     object
Year       int64
dtype: object

In [21]:
movies.columns # variables

Index(['Rank', 'Title', 'Studio', 'Gross', 'Year'], dtype='object')